File: FirebaseDatabase/Sources/Utilities/FNextPushId.m
/*
 * Copyright 2017 Google
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#import "FirebaseDatabase/Sources/Utilities/FNextPushId.h"
#import "FirebaseDatabase/Sources/Utilities/FUtilities.h"
#import "FirebaseDatabase/Sources/Utilities/FValidation.h"

static NSString *const PUSH_CHARS =
    @"-0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ_abcdefghijklmnopqrstuvwxyz";

static NSString *const MIN_PUSH_CHAR = @" ";

static NSString *const MAX_PUSH_CHAR = @"\uFFFF";

static NSInteger const MAX_KEY_LEN = 786;

static unichar const LOW_SURROGATE_PAIR_START = 0xDC00;
static unichar const LOW_SURROGATE_PAIR_END = 0xDFFF;
static unichar const HIGH_SURROGATE_PAIR_START = 0xD800;
static unichar const HIGH_SURROGATE_PAIR_END = 0xDBFF;

@implementation FNextPushId

+ (NSString *)get:(NSTimeInterval)currentTime {
    static long long lastPushTime = 0;
    static int lastRandChars[12];

    long long now = (long long)(currentTime * 1000);

    BOOL duplicateTime = now == lastPushTime;
    lastPushTime = now;

    unichar timeStampChars[8];
    for (int i = 7; i >= 0; i--) {
        timeStampChars[i] = [PUSH_CHARS characterAtIndex:(now % 64)];
        now = (long long)floor(now / 64);
    }

    NSMutableString *id = [[NSMutableString alloc] init];
    [id appendString:[NSString stringWithCharacters:timeStampChars length:8]];

    if (!duplicateTime) {
        for (int i = 0; i < 12; i++) {
            lastRandChars[i] = (int)floor(arc4random() % 64);
        }
    } else {
        int i = 0;
        for (i = 11; i >= 0 && lastRandChars[i] == 63; i--) {
            lastRandChars[i] = 0;
        }
        lastRandChars[i]++;
    }

    for (int i = 0; i < 12; i++) {
        [id appendFormat:@"%C", [PUSH_CHARS characterAtIndex:lastRandChars[i]]];
    }

    return [NSString stringWithString:id];
}

+ (NSString *)from:(NSString *)fn successor:(NSString *_Nonnull)key {
    [FValidation validateFrom:fn validKey:key];
    NSInteger keyAsInt;
    if ([FUtilities tryParseString:key asInt:&keyAsInt]) {
        if (keyAsInt == [FUtilities int32max]) {
            return MIN_PUSH_CHAR;
        }
        return [NSString stringWithFormat:@"%ld", (long)keyAsInt + 1];
    }
    NSMutableString *next = [NSMutableString stringWithString:key];
    if ([next length] < MAX_KEY_LEN) {
        [next insertString:MIN_PUSH_CHAR atIndex:[key length]];
        return next;
    }

    long i = [next length] - 1;
    while (i >= 0) {
        if ([next characterAtIndex:i] != [MAX_PUSH_CHAR characterAtIndex:0]) {
            break;
        }
        --i;
    }

    // `nextAfter` was called on the largest possible key, so return the
    // maxName, which sorts larger than all keys.
    if (i == -1) {
        return [FUtilities maxName];
    }

    unichar character = [next characterAtIndex:i];
    unichar plusOne = character + 1;
    BOOL removePreviousCharacter = NO;
    BOOL replaceWithLowestSurrogatePair = NO;
    switch (plusOne) {
    case 0x23: // '#'
    case 0x24: // '$'
        plusOne = 0x25;
        break;

    case 0x2E: // '.'
    case 0x2F: // '/'
        plusOne = 0x30;
        break;

    case 0x5B: // '['
        plusOne = 0x5C;
        break;

    case 0x5D: // ']'
        plusOne = 0x5E;
        break;

    case 0x7F: // control character: del
        plusOne = 0x80;
        break;

    case HIGH_SURROGATE_PAIR_START: // 0xD800
        // We added one to 0xD7FF and entered surrogate pair zone
        // Replace with the lowest surrogate pair here
        replaceWithLowestSurrogatePair = YES;

    case LOW_SURROGATE_PAIR_END + 1: // 0xE000
        // If the previous character is the highest surrogate value
        // then we increment to the value 0xE000 by replacing the surrogate
        // pair by the single value 0xE000 (the value of plusOne)
        // Otherwise we increment the high surrogate value and set the low
        // surrogate value to the lowest.
        if (i == 0) {
            // Error, encountered low surrogate without matching high surrogate
        } else {
            unichar high = [next characterAtIndex:i - 1];
            if (high == HIGH_SURROGATE_PAIR_END) { /* highest value for the high
                                                      part of the pair */
                // Replace pair with 0xE000 (the value of plusOne)
                removePreviousCharacter = YES;
            } else {
                high += 1;
                NSString *highStr = [NSString stringWithFormat:@"%C", high];

                [next replaceCharactersInRange:NSMakeRange(i - 1, i)
                                    withString:highStr];
                plusOne = LOW_SURROGATE_PAIR_START; /* lowest value for the low
                                                       part of the pair */
            }
        }
        break;
    }

    NSString *sourcePlusOne =
        replaceWithLowestSurrogatePair
            ? [NSString stringWithFormat:@"%C%C", HIGH_SURROGATE_PAIR_START,
                                         LOW_SURROGATE_PAIR_START]
            : [NSString stringWithFormat:@"%C", plusOne];

    NSInteger replaceLocation = i;
    NSInteger replaceLength = 1;
    if (removePreviousCharacter) {
        --replaceLocation;
        ++replaceLength;
    }

    [next replaceCharactersInRange:NSMakeRange(replaceLocation, replaceLength)
                        withString:sourcePlusOne];
    NSInteger length = i + 1;
    if (removePreviousCharacter) {
        --length;
    } else if (replaceWithLowestSurrogatePair) {
        ++length;
    }
    return [next substringWithRange:NSMakeRange(0, length)];
}

+ (NSString *)from:(NSString *)fn predecessor:(NSString *_Nonnull)key {
    [FValidation validateFrom:fn validKey:key];
    NSInteger keyAsInt;
    if ([FUtilities tryParseString:key asInt:&keyAsInt]) {
        if (keyAsInt == [FUtilities int32min]) {
            return [FUtilities minName];
        }
        return [NSString stringWithFormat:@"%ld", (long)keyAsInt - 1];
    }
    NSMutableString *next = [NSMutableString stringWithString:key];
    if ([next characterAtIndex:(next.length - 1)] ==
        [MIN_PUSH_CHAR characterAtIndex:0]) {
        if ([next length] == 1) {
            return
                [NSString stringWithFormat:@"%ld", (long)[FUtilities int32max]];
        }
        // If the last character is the smallest possible character, then the
        // next smallest string is the prefix of `key` without it.
        [next replaceCharactersInRange:NSMakeRange([next length] - 1, 1)
                            withString:@""];
        return next;
    }
    // Replace the last character with its immediate predecessor, and fill the
    // suffix of the key with MAX_PUSH_CHAR. This is the lexicographically
    // largest possible key smaller than `key`.
    NSUInteger i = next.length - 1;
    unichar character = [next characterAtIndex:i];
    unichar minusOne = character - 1;
    BOOL removePreviousCharacter = NO;
    BOOL replaceWithHighestSurrogatePair = NO;
    switch (minusOne) {
    // NOTE: We already handled the case of min char (0x20)
    case 0x23: // '#'
    case 0x24: // '$'
        minusOne = 0x22;
        break;

    case 0x2E: // '.'
    case 0x2F: // '/'
        minusOne = 0x2D;
        break;

    case 0x5B: // '['
        minusOne = 0x5A;
        break;

    case 0x5D: // ']'
        minusOne = 0x5C;
        break;

    case 0x7F: // control character: del
        minusOne = 0x7E;
        break;

    case LOW_SURROGATE_PAIR_END: // 0xDFFF
        // Previously we had 0xE000 which is a single utf16 character,
        // this needs to be replaced with the highest surrogate pair:
        replaceWithHighestSurrogatePair = YES;
        break;

    case HIGH_SURROGATE_PAIR_END: // 0xDBFF
        // If the previous character is the lowest high surrogate value
        // then we decrement to the non-surrogate value 0xD7FF by replacing the
        // surrogate pair by the single value 0xD7FF (HIGH_SURROGATE_PAIR_START
        // - 1) Otherwise we decrement the high surrogate value and set the low
        // surrogate value to the highest.
        if (i == 0) {
            // Error, found low surrogate without matching high surrogate
        } else {
            unichar high = [next characterAtIndex:i - 1];
            if (high == HIGH_SURROGATE_PAIR_START) { /* lowest value for the
                                                        high part of the pair */
                // Replace pair with single 0xD7FF value
                removePreviousCharacter = YES;
                minusOne = HIGH_SURROGATE_PAIR_START - 1;
            } else {
                high -= 1;
                NSString *highStr = [NSString stringWithFormat:@"%C", high];

                [next replaceCharactersInRange:NSMakeRange(i - 1, i)
                                    withString:highStr];
                minusOne = LOW_SURROGATE_PAIR_END; /* highest value for the low
                                                      part of the pair */
            }
        }
        break;
    }

    NSString *sourceMinusOne =
        replaceWithHighestSurrogatePair
            ? [NSString stringWithFormat:@"%C%C", HIGH_SURROGATE_PAIR_END,
                                         LOW_SURROGATE_PAIR_END]
            : [NSString stringWithFormat:@"%C", minusOne];

    NSInteger replaceLocation = i;
    NSInteger replaceLength = 1;
    if (removePreviousCharacter) {
        --replaceLocation;
        ++replaceLength;
    }

    [next replaceCharactersInRange:NSMakeRange(replaceLocation, replaceLength)
                        withString:sourceMinusOne];

    NSInteger length = i + 1;
    if (removePreviousCharacter) {
        --length;
    } else if (replaceWithHighestSurrogatePair) {
        ++length;
    }
    return [next stringByPaddingToLength:MAX_KEY_LEN
                              withString:MAX_PUSH_CHAR
                         startingAtIndex:0];
};

@end

File: FirebaseDatabaseSwiftCore/Sources/FNextPushId.swift
//
//  File.swift
//  
//
//  Created by Morten Bek Ditlevsen on 11/10/2021.
//

import Foundation

let PUSH_CHARS =
    "-0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ_abcdefghijklmnopqrstuvwxyz"

let MIN_PUSH_CHAR: Character = "-"

let MAX_PUSH_CHAR: Character = "z"

let MAX_KEY_LEN = 786

@objc public class FNextPushId: NSObject {
    private static var lastPushTime: Int64 = 0
    private static var lastRandChars: [UInt8] = Array<UInt8>(repeating: 0, count: 12)
    @objc public static func get(_ currentTime: TimeInterval) -> String {
        var now: Int64 = Int64(currentTime * 1000)
        let duplicateTime = now == lastPushTime
        lastPushTime = now

        var timeStampChars: String = ""
        for _ in 0 ..< 8 {
            let offset = Int(now % 64)
            let index = PUSH_CHARS.index(PUSH_CHARS.startIndex, offsetBy: offset)
            timeStampChars.append(PUSH_CHARS[index])
            now = now / 64
        }

        var id: String = String(timeStampChars.reversed())
        if !duplicateTime {
            for i in (0 ..< 12).reversed() {
                lastRandChars[i] = UInt8(arc4random() % 64)
            }
        } else {
            var j = 0
            for i in (0 ..< 12).reversed() {
                j = i
                guard lastRandChars[i] == 63 else { break }
                lastRandChars[i] = 0
            }
            lastRandChars[j] += 1
        }
        for i in 0 ..< 12 {
            let offset = Int(lastRandChars[i])
            let index = PUSH_CHARS.index(PUSH_CHARS.startIndex, offsetBy: offset)
            id.append(PUSH_CHARS[index])
        }

        return id
    }

    @objc public static func successor(_ key: String) -> String {
        var keyAsInt: Int = 0
        if tryParseStringToInt(key, integer: &keyAsInt) {
            if keyAsInt == Int(Int32.max) {
                return String(MIN_PUSH_CHAR)
            }
            return "\(keyAsInt + 1)"
        }
        if key.count < MAX_KEY_LEN {
            return key + String(MIN_PUSH_CHAR)
        }

        var next = key
        guard let index = next.lastIndex(where: { $0 != MAX_PUSH_CHAR }) else {
            // `successor` was called on the largest possible key, so return the
            // maxName, which sorts larger than all keys.
            return FUtilitiesSwift.maxName
        }
        let source = next[index]

        if let sourceIndex = PUSH_CHARS.firstIndex(of: source) {
            let sourcePlusOne = PUSH_CHARS[PUSH_CHARS.index(sourceIndex, offsetBy: 1)]
            next.replaceSubrange(index...index, with: String(sourcePlusOne))
            return String(next[next.startIndex...index])
        }
        fatalError("Existing implementation crashes if 'source' character is not included in PUSH_CHARS, so we may as well do the same...")
    }

    // `key` is assumed to be non-empty
    #warning("It would perhaps be good to assert this, or even to support this, because even though keys can't be empty, queryBefore for instance may make sence on an empty String...")
    // In this implementation 'key' is not assumed to be non-empty, but the key that comes
    // before the empty string is Int32.max
    @objc public static func predecessor(_ key: String) -> String {
        var keyAsInt: Int = 0
        if tryParseStringToInt(key, integer: &keyAsInt) {
            if keyAsInt == Int(Int32.min) {
                return FUtilitiesSwift.minName
            }
            return "\(keyAsInt - 1)"
        }
        if key.last == MIN_PUSH_CHAR {
            if key.count == 1 {
                return "\(Int32.max)"
            }
            // If the last character is the smallest possible character, then the
            // next smallest string is the prefix of `key` without it.
            return String(key.dropLast())
        }
        var next = key
        // Replace the last character with its immedate predecessor, and fill the
        // suffix of the key with MAX_PUSH_CHAR. This is the lexicographically
        // largest possible key smaller than `key`.
        guard let curr = next.last else {
            return "\(Int32.max)"
        }
        if let sourceIndex = PUSH_CHARS.firstIndex(of: curr) {
            let sourceMinusOne = PUSH_CHARS[PUSH_CHARS.index(sourceIndex, offsetBy: -1)]
            let index = next.index(before: next.endIndex)
            next.replaceSubrange(index...index, with: String(sourceMinusOne))
            return next.padding(toLength: MAX_KEY_LEN, withPad: String(MAX_PUSH_CHAR), startingAt: 0)
        }
        fatalError("Existing implementation crashes if 'curr' character is not included in PUSH_CHARS, so we may as well do the same...")

    }
}
